src/Razor/test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Cohost/CohostOnAutoInsertEndpointTest.cs

```diff
@@ -0,0 +1,275 @@
+
+        var tab = '\t';
+        var output = $$"""
+        @code {
+        {{tab}}void TestMethod()
+        {{tab}}{
+        {{tab}}{{tab}}$0
+        {{tab}}}
+        }
+        """;
+        await VerifyOnAutoInsertAsync(input, output, triggerCharacter: "\n", insertSpaces: false);
+    }
+
+    private async Task VerifyOnAutoInsertAsync(
+        string input,
+        string output,
+        string triggerCharacter,
+        string? delegatedResponseText = null,
+        bool insertSpaces = true,
+        int tabSize = 4,
+        bool formatOnType = true,
+        bool autoClosingTags = true,
+        bool expectResult = true)
+    {
+        var testCode = new TestCode(input);
+     
+        var document = CreateProjectAndRazorDocument(testCode.Text);
+        var sourceText = await document.GetTextAsync(DisposalToken);
+
+        var clientSettingsManager = new ClientSettingsManager([], null, null);
+        clientSettingsManager.Update(ClientAdvancedSettings.Default with { FormatOnType = formatOnType, AutoClosingTags = autoClosingTags });
+
+        IOnAutoInsertTriggerCharacterProvider[] onAutoInsertTriggerCharacterProviders = [
+            new RemoteAutoClosingTagOnAutoInsertProvider(),
+            new RemoteCloseTextTagOnAutoInsertProvider()];
+
+        VSInternalDocumentOnAutoInsertResponseItem? response = null;
+        if (delegatedResponseText is not null)
+        {
+            var start = sourceText.GetPosition(testCode.Position);
+            var end = start;
+            response = new VSInternalDocumentOnAutoInsertResponseItem()
+            {
+                TextEdit = new TextEdit() { NewText = delegatedResponseText, Range = new() { Start = start, End = end } },
+                TextEditFormat = InsertTextFormat.Snippet
+            };
+        }
+
+        var requestInvoker = new TestLSPRequestInvoker([(VSInternalMethods.OnAutoInsertName, response)]);
+
+        var endpoint = new CohostOnAutoInsertEndpoint(
+            RemoteServiceInvoker,
+            clientSettingsManager,
+            onAutoInsertTriggerCharacterProviders,
+            TestHtmlDocumentSynchronizer.Instance,
+            requestInvoker,
+            LoggerFactory);
+
+        var formattingOptions = new FormattingOptions()
+        {
+            InsertSpaces = insertSpaces,
+            TabSize = tabSize
+        };
+
+        var request = new VSInternalDocumentOnAutoInsertParams()
+        {
+            TextDocument = new TextDocumentIdentifier()
+            {
+                Uri = document.CreateUri()
+            },
+            Position = sourceText.GetPosition(testCode.Position),
+            Character = triggerCharacter,
+            Options = formattingOptions
+        };
+
+        var result = await endpoint.GetTestAccessor().HandleRequestAsync(request, document, DisposalToken);
+
+        if (expectResult)
+        {
+            Assert.NotNull(result);
+        }
+        else
+        {
+            Assert.Null(result);
```

Comment:
It seems a little onerous to have to specify the output exactly the same as the input, when the test expects nothing from the service. An early return here would remove that requirement, and then `output` could be made nullable, and then `expectResult` could be removed, and the check above this can just be a null check on the output.

I like this assert though, definitely keep that!